metadata/src/lib.rs

```diff
@@ -593,74 +605,93 @@ impl ConstantMetadata {
     signature_ty: u32,
-    /// The type of the outermost Extra enum.
-    extra_ty: u32,
     /// Which extrinsic versions are supported by this chain.
     supported_versions: Vec<u8>,
     /// The signed extensions in the order they appear in the extrinsic.
-    transaction_extensions: Vec<TransactionExtensionMetadata>,
-    /// Version of the transaction extensions.
-    // TODO [jsdw]: V16 metadata groups transaction extensions by version.
-    // need to work out what to do once there is more than one version to deal with.
-    transaction_extensions_version: u8,
+    transaction_extensions: Vec<TransactionExtensionMetadataInner>,
+    /// Different versions of transaction extensions can exist. Each version
+    /// is a u8 which corresponds to the indexes of the transaction extensions
+    /// seen in the above Vec, in order, that exist at that version.
+    transaction_extensions_by_version: BTreeMap<u8, Vec<u32>>,
 }
 
 impl ExtrinsicMetadata {
-    /// The type of the address that signs the extrinsic
-    pub fn address_ty(&self) -> u32 {
-        self.address_ty
-    }
-
-    /// The type of the outermost Call enum.
-    pub fn call_ty(&self) -> u32 {
-        self.call_ty
-    }
-    /// The type of the extrinsic's signature.
-    pub fn signature_ty(&self) -> u32 {
-        self.signature_ty
-    }
-    /// The type of the outermost Extra enum.
-    pub fn extra_ty(&self) -> u32 {
-        self.extra_ty
-    }
-
     /// Which extrinsic versions are supported.
     pub fn supported_versions(&self) -> &[u8] {
         &self.supported_versions
     }
 
     /// The extra/additional information associated with the extrinsic.
-    pub fn transaction_extensions(&self) -> &[TransactionExtensionMetadata] {
-        &self.transaction_extensions
+    pub fn transaction_extensions_by_version(
+        &self,
+        version: u8,
+    ) -> Option<impl Iterator<Item = TransactionExtensionMetadata<'_>>> {
+        let extension_indexes = self.transaction_extensions_by_version.get(&version)?;
+        let iter = extension_indexes.iter().map(|index| {
+            let tx_metadata = self
+                .transaction_extensions
+                .get(*index as usize)
+                .expect("transaction extension should exist if index is in transaction_extensions_by_version");
+
+            TransactionExtensionMetadata {
+                identifier: &tx_metadata.identifier,
+                extra_ty: tx_metadata.extra_ty,
+                additional_ty: tx_metadata.additional_ty,
+            }
+        });
+
+        Some(iter)
     }
 
-    /// Which version are these transaction extensions?
-    pub fn transaction_extensions_version(&self) -> u8 {
-        self.transaction_extensions_version
+    /// When constructing a v5 extrinsic, use this transaction extensions version.
+    pub fn transaction_extension_version_to_use_for_encoding(&self) -> u8 {
+        *self.transaction_extensions_by_version.keys().max().unwrap()
```

Comment:
Silly comment: maybe add an expect that `transaction_extensions_by_version` must contain at least one version

Reply:
Done!